src/_P103_Atlas_EZO_pH_ORP_EC.ino

```diff
@@ -0,0 +1,641 @@
+      char voltagedata[32];
+      status = _P103_send_I2C_command(I2Cchoice, F("Status"), voltagedata);
+
+
+      if (status) {
+        String sensorString(sensordata);
+        String voltage(voltagedata);
+        int    pos = voltage.lastIndexOf(',');
+        UserVar[event->BaseVarIndex]     = sensorString.toFloat();
+        UserVar[event->BaseVarIndex + 1] = voltage.substring(pos + 1).toFloat();
+      }
+      else {
+        UserVar[event->BaseVarIndex]     = -1;
+        UserVar[event->BaseVarIndex + 1] = -1;
+      }
+
+      success = true;
+      break;
+    }
+  }
+  return success;
+}
+
+// Call this function with two char arrays, one containing the command
+// The other containing an allocatted char array for answer
+// Returns true on success, false otherwise
+
+bool _P103_send_I2C_command(uint8_t I2Caddress, const String& cmd, char *sensordata) {
+  uint16_t sensor_bytes_received = 0;
+
+  byte error;
+  byte i2c_response_code = 0;
+  byte in_char           = 0;
+
+  String log = F("> cmd = ");
+          log += String(cmd);
+          addLog(LOG_LEVEL_DEBUG, log);
+
+  addLog(LOG_LEVEL_DEBUG, String(cmd));
+  Wire.beginTransmission(I2Caddress);
+  Wire.write(cmd.c_str());
+  error = Wire.endTransmission();
+
+  if (error != 0) {
+    // addLog(LOG_LEVEL_ERROR, error);
+    addLog(LOG_LEVEL_ERROR, F("Wire.endTransmission() returns error: Check Atlas shield, pH, ORP and EC are supported."));
+    return false;
+  }
+
+  // don't read answer if we want to go to sleep
+  if (cmd.substring(0, 5).equalsIgnoreCase(F("Sleep"))) {
+    return true;
+  }
+
+  i2c_response_code = 254;
+
+  while (i2c_response_code == 254) { // in case the command takes longer to process, we keep looping here until we get a success or an error
+    if  (
+      (((cmd[0] == 'r') || (cmd[0] == 'R')) && (cmd[1] == '\0'))
+      ||
+      (cmd.substring(0, 3).equalsIgnoreCase(F("cal")) && !cmd.substring(0, 5).equalsIgnoreCase(F("Cal,?")))
+      )
+    {
+      // FIXME TD-er: No way we will wait this long.
+      //delay(900);
+    }
+    else {
+      // FIXME TD-er: No way we will wait this long.
+      //delay(300);
+    }
+
+    Wire.requestFrom(I2Caddress, (uint8_t)32); // call the circuit and request 32 bytes (this is more then we need).
+    i2c_response_code = Wire.read();           // read response code
+
+    while (Wire.available()) {                 // read response
+      in_char = Wire.read();
+
+      if (in_char == 0) {                      // if we receive a null caracter, we're done
+        while (Wire.available()) {             // purge the data line if needed
+          Wire.read();
+        }
+
+        break;                                       // exit the while loop.
+      }
+      else {
+        sensordata[sensor_bytes_received] = in_char; // load this byte into our array.
+        sensor_bytes_received++;
+      }
+    }
+    sensordata[sensor_bytes_received] = '\0';
```

Comment:
This zero-termination is only performed if we enter `while (i2c_response_code == 254) {` at least once.
So perhaps best to set at least the first character to 0 also, just to make sure the `sensordata` array is zero-terminated.

Reply:
ok

Reply:
I took a look at this. We set `i2c_response_code = 254` directly before the while directive so we will enter there every time. I will initialise `sensordata[32] = {0};` every time in the code when we use it. So I think this should be OK. Di you agree or do you hav any other suggestions?

Reply:
In that while loop you do set the "next" char to 0, so better initialize the first char to 0 or else you may still have garbage but at least the ESP will not crash as it tries to access out of array bounds. :)

Also make sure `sensor_bytes_received` never exceeds the length of the array.

Reply:
OK i did this. know I initlaize the first character to 0 and check the length.